### DISCUSSION TITLE
I need some help with a simple drag and drop application.

### DISCUSSION THREAD

**Discussion by Z1-Gamer** [AllenDang/giu#682] [Q&A] [answered]
Hi! I am trying to make a simple application in which I want to drag and drop an image file from windows explorer into an image widget inside the program.

Here are some relevant snippets of what I have so far and the problem will be explained below.

		giu.Row(
                giu.ImageWithFile(Image1Url).OnClick(onClickMe).Size(240, Image1sizeX),
	            giu.Custom(func() {
			if giu.IsItemHovered() {
				Image1Hovered = true
			} else {
				Image1Hovered = false
			}

In the above code I use a new bool variable to know if the image widget is hovered.

Down in my main function I have the callback for when something is dropped in the application.
`wnd.SetDropCallback(onDrop)`

After the drop call, in the onDrop function i use this

            if Image1Hovered {
                ImageFile1, err := os.Open(names[0])

which practically checks if my image widget was hovered over at the moment the file was dropped in the program, and from there I can proceed and do what I need. This almost works perfectly.

Here is the problem. The isItemHovered only returns true if the program window is on top of all others. That means if for example I have a folder and the program open side by side, and I have opened the folder last, when I start to drag&drop a file, the explorer window is the one "on top" and the isItemHovered reports false even though the program window is completely visible and I can see where I am dropping the file.
Now if I click on the program window while I have them both side by side, and then directly drag&drop a file, the program window is the one on top and everything works as intended.

So my question is how can I make it work properly (in both case above instead only during the second one)?
I am new to this framework so i apologize if my question is dumb. 

I will be grateful for any help, thanks!

**Comment by gucio321** [comment]
Well this is an interesting question.
I have two ideas on top of my head:
First one requirrs to disable mechanic called power saving mode but as far as I know it isn't possible (and it'd be quite ineffective)
Another way is to use giu.Update to enforce ui refreshing.
- create another variable and temporary store image path there
- in your drop callback call giu.Update

**Reply by Z1-Gamer** [reply]
Thanks for your quick reply. Unfortunately I don't quite understand.
Do you mean to add the giu.update in the onDrop funcion? If so there is already a call to giu.update at the end of onDrop.
Just to be sure I also tried adding it on the line just above

           if Image1Hovered {
               ImageFile1, err := os.Open(names[0])

so it became like this, I will just paste my full onDrop function (a little long).


     func onDrop(names []string) {
		giu.Update()
		if Image1Hovered {
			ImageFile1, err := os.Open(names[0])
			if err != nil {
				log.Fatal(err)
			}
			Image1, _, err2 := image.DecodeConfig(ImageFile1)
			if err2 != nil {
				log.Fatal(err2)
			}

			if Image1.Width < Image1.Height {
				Image1Ratio = float32(Image1.Height) / float32(Image1.Width)
				Image1sizeX = 240 * Image1Ratio
			} else {
				Image1Ratio = float32(Image1.Width) / float32(Image1.Height)
				Image1sizeX = 240 / Image1Ratio
			}
			Image1Url = names[0]
		}

		if Image2Hovered {
			ImageFile2, err := os.Open(names[0])
			if err != nil {
				log.Fatal(err)
			}
			Image2, _, err2 := image.DecodeConfig(ImageFile2)
			if err2 != nil {
				log.Fatal(err2)
			}

			if Image2.Width < Image2.Height {
				Image2Ratio = float32(Image2.Height) / float32(Image2.Width)
				Image2sizeX = 240 * Image2Ratio
			} else {
				Image2Ratio = float32(Image2.Width) / float32(Image2.Height)
				Image2sizeX = 240 / Image2Ratio
			}
			Image2Url = names[0]
		}

		giu.Update()

	}


But that did nothing as well. The problem is that in the onDrop funtion, when a file is dropped the program never enters "if Image1Hovered {" because it is false (when the program is not on top).

If I use giu.MasterWindowFlagsFloating as a flag in my master window, the application is always on top so it always works, but I just don't want to force it to be always on top of everything if I can avoid it.

**Reply by gucio321** [reply]
Well so now you need to move your whole ondrop function into this onHover trigger and pass paths list there, this will execute your code only when image is hovered (I'm sorry if i write a bit enigmatic, I'd post a full code but unfortunately there is no PC around 😁)

**Reply by gucio321** [reply]
Something like
Ondrop:
```go
globalPaths=names
giu.Update()
```
And the rest of your code put in this hover detector below image widget

**Reply by Z1-Gamer** [reply]
No matter what I put under `if giu.IsItemHovered() {` in the full code below, the event never gets triggered when the window is not on top, even when I am hovering over the widget. So no code will ever be executed there when the window is not on top.


        giu.Row(
            giu.ImageWithFile(Image1Url).OnClick(onClickMe).Size(240, Image1sizeX),
            giu.Custom(func() {
		if giu.IsItemHovered() {
			Image1Hovered = true
		} else {
			Image1Hovered = false
		}`


One another note, i was looking through the documentation and i found this:

> // HoveredFlagsAllowWhenBlockedByActiveItem Return true even if an active item is blocking access to this item/window.
	// Useful for Drag and Drop patterns.
	[HoveredFlags](https://pkg.go.dev/github.com/AllenDang/giu#HoveredFlags)AllowWhenBlockedByActiveItem HoveredFlags = imgui.HoveredFlagsAllowWhenBlockedByActiveItem

Could this help me in some way? I am not sure where to set these flags.

**Reply by gucio321** [reply]
> Could this help me in some way? I am not sure where to set these flags

No because when window is not focused power saving mode causes that interface is not refreshed are all so no event can happen

> No matter what I put under if giu.IsItemHovered() { in the full code below, the event never gets triggered when the window is not on top, even when I am hovering over the widget. So no code will ever be executed there when the window is not on top.

This is because ui is not refreshed so no event can be detected. The trick is to call giu.Uptate when you want to check for this event so when OnDrop happens (ondrop isn't giu but glfw event so it is not affected by power saving mode)

**Reply by Z1-Gamer** [reply]
If what you say is true than with this code it should all work:

    func onDrop(names []string) {
         giu.Update()
	     if Image1Hovered {
		ImageFile1, err := os.Open(names[0])
		if err != nil {
			log.Fatal(err)
	        }

But i just tested it and it does not. calling giu.update does enable the trigger of the onHover event (when hovering).

**Reply by Z1-Gamer** [reply, marked as answer]
I found a way to make it work but there is now another problem....
So i followed some of your previous advice and here is how i made it currently.

      func onDrop(names []string) {
           JustDropped = true
	       ImageURL = names[0]
	       giu.Update()
      }
This is the whole of my onDrop function now.

And here is the full code where the widget is:

`   

        giu.ImageWithFile(Image1Url).OnClick(onClickMe).Size(240, Image1sizeX),
	        giu.Custom(func() {
		    if giu.IsItemHovered() && JustDropped {
		        JustDropped = false
			ImageFile1, err := os.Open(ImageURL)
			if err != nil {
				log.Fatal(err)
			}
			Image1, _, err2 := image.DecodeConfig(ImageFile1)
			if err2 != nil {
				log.Fatal(err2)
			}

			if Image1.Width < Image1.Height {
				Image1Ratio = float32(Image1.Height) / float32(Image1.Width)
				Image1sizeX = 240 * Image1Ratio
			} else {
				Image1Ratio = float32(Image1.Width) / float32(Image1.Height)
				Image1sizeX = 240 / Image1Ratio
			}
			Image1Url = ImageURL
		}

Now this works, when the window is not on top and I drop an image in the widget it is shown instantly (as intended).

The problem is now in the reverse case. When the window is on top and I drop an image in the widget, it does not immediately get placed inside the widget, only after I move the mouse cursor the image gets shown.
Any way to fix that?

While previously with the old code it worked perfectly when the window was on top.
So in one case i have code that works perfectly only when the window is on top, and now in this new case i have code that works perfectly only when the window is not on top. Gotta love programming :D.

**Reply by Z1-Gamer** [reply]
Well i finally made it work. just added `giu.Update()` right after i set the image with  `Image1Url = ImageURL` which is at the end of the If statement in the widget.

Thanks for your help.

**Reply by gucio321** [reply]
No problem 😁
